src/types/string.typ

```diff
@@ -9,6 +9,35 @@
 /// -> schema
 #let string = base-type.with(name: "string", types: (str,))
 
+#let string(
+  assertions: (),
+  min: none,
+  max: none,
+  ..args,
+) = {
+
+  assert-positive-type(min, types: (int,), name: "Minimum length")
+  assert-positive-type(max, types: (int,), name: "Maximum length")
+
+  base-type(name: "string", types: (str,), ..args) + (
+    min: min,
+    max: max,
+    assertions: (
+      (
+        precondition: "min",
+        condition: (self, it) => it.len() >= self.min,
+        message: (self, it) => "Length must be at least " + str(self.min),
+      ),
+      (
+        precondition: "max",
+        condition: (self, it) => it.len() <= self.max,
+        message: (self, it) => "Length must be at most " + str(self.max),
+      ),
+      ..assertions,
+    ),
```

Comment:
The original reason for these to be removed was the duplication, and I think we can probably extract these preconditions into some basic factories which create these, the weak type system works to works advantage.

One which compares values as is without calling `.len()` and these which work on collections could simply call wrapped the lambda from the original and call `.len()` beforehand.

Reply:
This is a good point. The corner I've painted myself into is that I would like to store the righthand side of the comparison on the schema object, so that parent containers might still be able to change their behaviour. There isn't a such type yet, but keeping I'd still like to keep that avenue open.

I'll keep having a think and see if there's a better way

Reply:
Not all duplication is bad too, up to you what you see as more convenient.